Lib/asyncio/base_events.py

```diff
@@ -155,6 +156,63 @@ def _run_until_complete_cb(fut):
     futures._get_loop(fut).stop()
 
 
+
+class _SendfileProtocol(protocols.Protocol):
+    def __init__(self, transp):
+        # transport should be _FlowControlMixin instance
+        self._transport = transp
+        self._proto = transp.get_protocol()
+        self._should_resume_reading = transp.is_reading()
+        self._should_resume_writing = transp._protocol_paused
+        transp.pause_reading()
+        transp.set_protocol(self)
+        if self._should_resume_writing:
+            self._paused = self._transport._loop.create_future()
+        else:
+            self._paused = None
+
+    def connection_made(self, transport):
+        raise RuntimeError("Invalid state, "
+                           "connection should have been established already.")
+
+    def connection_lost(self, exc):
+        if self._paused is not None:
+            # Never happens if peer disconnects after sending the whole content
+            # Thus disconnection is always an exception from user perspective
+            if exc is None:
+                self._paused.set_exception(
+                    ConnectionError("Connection is closed by peer"))
+            else:
+                self._paused.set_exception(exc)
+        self._proto.connection_lost(exc)
+
+    def pause_writing(self):
+        if self._paused is not None:
+            return
+        self._paused = self._transport._loop.create_future()
+
+    def resume_writing(self):
+        if self._paused is None:
+            return
+        self._paused.set_result(False)
+        self._paused = None
+
+    def data_received(self, data):
+        raise RuntimeError("Invalid state, reading should be paused")
+
+    def eof_received(self):
+        raise RuntimeError("Invalid state, reading should be paused")
+
+    async def restore(self):
+        self._transport.set_protocol(self._proto)
+        if self._should_resume_reading:
+            self._transport.resume_reading()
+        if self._paused is not None:
+            await self._paused
```

Comment:
This code is complex so I want it to be crystal clear.  Please add a comment before the await explaining what exactly we are waiting for.

Reply:
Something akin to:

```python
if self._can_write is not None:
    # Wait until the transport's buffer is drained and
    # writing can be resumed.  This is important to
    # allow the replaced protocol to resume its operations.
    await self._can_write
    self._can_write = None
```